### DISCUSSION TITLE
audio injector octo (cs42448) running!

### DISCUSSION THREAD

**Discussion by nickverlinden** [rsta2/circle#324] [Development]
Hi!

I got the octo running by creating a class for the cs42448 which initialised the chip over i2c and is able to start and configure the fpga clock on the board. The cs42448 is set up to send 8 channel TDM instead of regular 2 channel i2s. Then I set up an i2s TXRX device, and created a drain that just reads the input and dumps to the output. I did these modifications in the 42-i2sinput sample. 

If anybody wants to use the octo, I could not for the life of me get it to work reliably when it was not on top of the pi connected as a hat. Because the octo was at another place in my enclosure, i first tried to connect all the required pins using dupont cables, but that didn't work out. Even with very short cables. The i2c communication was very flakey. The only way it worked was to put it on top of the pi.

I only had to do some minor modifications to circle to make it work:

- **EDIT: this was not required to get the stereo in/out working. I changed this while trying some stuff to get all the 8 channels to work.** soundbasedevice.cpp: remove the asserts checking for max channels `assert (1 <= nChannels && nChannels <= 2);`
- soundbasedevice.h: change `#define SOUND_HW_CHANNELS	2` to `#define SOUND_HW_CHANNELS	8`

The drain is working, i hear what appears to be a perfectly correct stereo output from the input (my phone playing music). So i'm already very happy! However I want to get all of the 8 channel out, and 6 channel in working. I tried poking around, changing some other channel related stuff in i2ssoundbasedevice, but I don't really know what i'm doing. Right now, with the modifications above, if i plug in some audio on channel 3, and listen to the output of channel 3 i hear low rumble noise that gets louder when there's a beat in the music, but only on channel 3. When i change the cable to input 4, the same applies for output 4. So, the same for channel 4, 5 and 6. Note that if nothing is plugged into channel 3, there is no noise. So something about the 8 channels is definitely working. It sounds like there are missing a lot of bits per sample. I tried changing the WRITE_CHANNELS to 8 already (and removing the asserts in soundbasedevice.cpp), but the result is kind of weird. The output on channels 1 & 2 becomes the same (so the music sounds mono) but no distortion or anything else, and the other channels still experience the same low rumble noise when testing the input (same as before).

this is the drain code:
```c
// copy sound data
	m_Logger.Write (FromKernel, LogNotice, "draining sound from input directly to output");
	for (unsigned nCount = 0; 1; nCount++)
	{
		u8 Buffer[TYPE_SIZE*WRITE_CHANNELS*4096];
		int nBytes = m_pSound->Read (Buffer, sizeof Buffer);
		if (nBytes > 0)
		{
			int nResult = m_pSound->Write (Buffer, nBytes);
			if (nResult != nBytes)
			{
				m_Logger.Write (FromKernel, LogWarning, "Sound data dropped");
			}

#ifdef ENABLE_RECORDER
			nResult = pRecorder->Write (Buffer, nBytes);
			if (nResult != nBytes)
			{
				m_Logger.Write (FromKernel, LogWarning,
						"Sound data dropped (recorder)");
			}
#endif
		}

		m_Screen.Rotor (0, nCount);

		m_Scheduler.Yield ();
	}
```

This is in the config.h:
```c
#define SAMPLE_RATE	44100		// overall system clock

#define WRITE_FORMAT	2		// 0: 8-bit unsigned, 1: 16-bit signed, 2: 24-bit signed
#define WRITE_CHANNELS	2		// 1: Mono, 2: Stereo

#define VOLUME		0.5		// [0.0, 1.0]

#define QUEUE_SIZE_MSECS 100		// size of the sound queue in milliseconds duration
#define CHUNK_SIZE	(384 * 10)	// number of samples, written to sound device at once

#define DAC_I2C_ADDRESS	0		// I2C slave address of the DAC (0 for auto probing)
```

Is there anybody who could help me figure it out?

@rsta2 I see your making changes to the sound api on the development branch. I need to clean up the cs42448 class, because now it's hacked together. I suppose when i want to contribute the code, i'll hold of refactoring until the new soundbasedevice api is finalised and contribute then?

**Comment by rsta2** [comment]
Well done! Unfortunately I'm not familiar with the Octo Audio Injector and cannot say, if the I2S sound driver of Circle is able to work with it with 8 channels. That means perhaps 2 channels will work on the hardware side, but not 8? I guess [this code](https://github.com/phorton1/circle-prh/blob/master/audio) supports 8 channels. But perhaps you've already done the necessary update to the driver.

What I can say is, that the class `CSoundBaseDevice` is hard-coded for Mono-/Stereo-operation on the application side. That means it is not enough to increase the number of channels define to 8 and remove the asserts. You must go through the whole class implementation and see, where the number of channels is handled. Please note that there are two sound data formats, which this class handles, the hardware format and the write/read format for the application. The class converts between these formats (WRITE -> HW on output, HW -> READ on input). That is why changing the parameter `WRITE_FORMAT` in the *config.h* of the sample program does not have an effect. It simply changes the format, which is generated by the application, but not the hardware format, which is presented to the hardware.

Unfortunately I'm currently do not have much time left and would like to concentrate on the USB audio support, I'm working on. I thought about, if it is already the time to go to a variable number of channels, but this would create another much more workload and there is already enough. My goal is to create a new Circle release still this year, perhaps we can go to more channels later next year. I would expect your contribution of a CS42448 driver then.

**Reply by nickverlinden** [reply]
Yeah, the repo you sent me is the one from the vGuitar rig where I was looking at.

I think I would be able to figure out the channel multiplexing stuff myself, but it's all the DMA controller code inside i2ssoundbasedevice.cpp that I have a difficulty understanding. Do you think i'll have to touch that as well?

**Comment by nickverlinden** [comment]
Good news, got it to work. All 6ch in and 8ch out. The modification is once again super simple. I didn't even have to touch any of the DMA voodoo.

@rsta2 I guess i'll wait until the new circle version to clean this all up and contribute. Good luck in implementing usb audio, looking forward to using that! 

For other people trying to use the octo:

As a quick hack i changed the following sections in SoundBaseDevice.c:
```c
// around line 253, inside function 'int CSoundBaseDevice::Write (const void *pBuffer, size_t nCount)'

if (m_nWriteChannels == 2)
				{
					ConvertSoundFormat (Frame+m_nHWSampleSize, pBuffer8);
					pBuffer8 += m_nWriteSampleSize;
				}
				else if (m_nWriteChannels == 8)
				{
					ConvertSoundFormat (Frame+m_nHWSampleSize, pBuffer8);
					pBuffer8 += m_nWriteSampleSize;

					ConvertSoundFormat (Frame+(m_nHWSampleSize*2), pBuffer8);
					pBuffer8 += m_nWriteSampleSize;

					ConvertSoundFormat (Frame+(m_nHWSampleSize*3), pBuffer8);
					pBuffer8 += m_nWriteSampleSize;

					ConvertSoundFormat (Frame+(m_nHWSampleSize*4), pBuffer8);
					pBuffer8 += m_nWriteSampleSize;

					ConvertSoundFormat (Frame+(m_nHWSampleSize*5), pBuffer8);
					pBuffer8 += m_nWriteSampleSize;

					ConvertSoundFormat (Frame+(m_nHWSampleSize*6), pBuffer8);
					pBuffer8 += m_nWriteSampleSize;

					ConvertSoundFormat (Frame+(m_nHWSampleSize*7), pBuffer8);
					pBuffer8 += m_nWriteSampleSize;
				}
				else
				{
					memcpy (Frame+m_nHWSampleSize, Frame, m_nHWSampleSize);
				}
```
```c
// inside function 'int CSoundBaseDevice::Read (void *pBuffer, size_t nCount)'

if (m_nReadChannels == 2)
			{
				ConvertReadSoundFormat (pBuffer8, Frame);
				pBuffer8 += m_nReadSampleSize;

				ConvertReadSoundFormat (pBuffer8, Frame+m_nHWSampleSize);
				pBuffer8 += m_nReadSampleSize;
			}
			else if (m_nReadChannels == 8)
			{
				ConvertReadSoundFormat (pBuffer8, Frame);
				pBuffer8 += m_nReadSampleSize;

				ConvertReadSoundFormat (pBuffer8, Frame+m_nHWSampleSize);
				pBuffer8 += m_nReadSampleSize;

				ConvertReadSoundFormat (pBuffer8, Frame+(m_nHWSampleSize*2));
				pBuffer8 += m_nReadSampleSize;

				ConvertReadSoundFormat (pBuffer8, Frame+(m_nHWSampleSize*3));
				pBuffer8 += m_nReadSampleSize;

				ConvertReadSoundFormat (pBuffer8, Frame+(m_nHWSampleSize*4));
				pBuffer8 += m_nReadSampleSize;

				ConvertReadSoundFormat (pBuffer8, Frame+(m_nHWSampleSize*5));
				pBuffer8 += m_nReadSampleSize;

				ConvertReadSoundFormat (pBuffer8, Frame+(m_nHWSampleSize*6));
				pBuffer8 += m_nReadSampleSize;

				ConvertReadSoundFormat (pBuffer8, Frame+(m_nHWSampleSize*7));
				pBuffer8 += m_nReadSampleSize;
			}
			else
			{
				if (m_bLeftChannel)
				{
					ConvertReadSoundFormat (pBuffer8, Frame);
				}
				else
				{
					ConvertReadSoundFormat (pBuffer8, Frame+m_nHWSampleSize);
				}

				pBuffer8 += m_nReadSampleSize;
			}
```

Also remove the asserts checking for the number of channels as stated in the first comment in this discussion.

Then in the config.h of sample 34-soundbasedevice, change:
```c
#define WRITE_CHANNELS	8
```

**Reply by rsta2** [reply]
Great that it's working! Yes, we should wait for the release after the next for Octo support. Thanks for the wishes!
